v2/lints/cabf_br/lint_san_dns_name_onion_invalid.go

```diff
@@ -0,0 +1,145 @@
+requires the `.onion` name to be well-formed, even prior to RFC 7686.
+
+See also https://github.com/cabforum/documents/issues/191
+*******************************************************************/
+
+package cabf_br
+
+import (
+	"encoding/base32"
+	"fmt"
+	"strings"
+
+	"github.com/zmap/zcrypto/x509"
+	"github.com/zmap/zlint/v2/lint"
+	"github.com/zmap/zlint/v2/util"
+)
+
+var (
+	// Per 2.4 of Rendezvous v2:
+	// Valid onion addresses contain 16 characters in a-z2-7 plus ".onion"
+	onionV2Len = 16
+
+	// Per 1.2 of Rendezvous v3:
+	// A hidden service's name is its long term master identity key.  This is
+	// encoded as a hostname by encoding the entire key in Base 32, including a
+	// version byte and a checksum, and then appending the string ".onion" at the
+	// end. The result is a 56-character domain name.
+	onionV3Len = 56
+)
+
+type onionNotValid struct{}
+
+func (l *onionNotValid) Initialize() error {
+	return nil
+}
+
+// CheckApplies returns true if the certificate contains one or more subject
+// names ending in `.onion`.
+func (l *onionNotValid) CheckApplies(c *x509.Certificate) bool {
+	// TODO(sleevi): This should also be extended to support nameConstraints
+	// in the future.
+	return util.CertificateSubjInTLD(c, util.OnionTLD)
+}
+
+// Execute will lint the provided certificate. A lint.Error lint.LintResult will
+// be returned if:
+//
+//  1) The certificate contains a Tor Rendezvous Spec v2 address and is not an
+//     EV certificate (BRs: Appendix C).
+//  2) The certificate contains a `.onion` subject name/SAN that is neither a
+//     Rendezvous Spec v2 or v3 address.
+func (l *onionNotValid) Execute(c *x509.Certificate) *lint.LintResult {
+	for _, subj := range append(c.DNSNames, c.Subject.CommonName) {
+		if !strings.HasSuffix(subj, util.OnionTLD) {
+			continue
+		}
+		labels := strings.Split(subj, ".")
+		if len(labels) < 2 {
+			return &lint.LintResult{
+				Status: lint.Error,
+				Details: fmt.Sprintf("certificate contained a %s domain with too "+
+					"few labels: %q", util.OnionTLD, subj),
+			}
+		}
+		onionDomain := labels[len(labels)-2]
+		if len(onionDomain) == onionV2Len {
+			// Onion v2 address. These are only permitted for EV, per BRs Appendix C.
+			if !util.IsEV(c.PolicyIdentifiers) {
+				return &lint.LintResult{
+					Status: lint.Error,
+					Details: fmt.Sprintf("%q is a v2 address, but the certificate is not "+
+						"EV", subj),
+				}
+			}
+		} else if len(onionDomain) == onionV3Len {
+			// Onion v3 address. Permitted for all certificates by CA/Browser Forum
+			// Ballot SC27.
+		} else {
+			return &lint.LintResult{
+				Status:  lint.Error,
+				Details: fmt.Sprintf("%q is not a v2 or v3 Tor address", subj),
+			}
+		}
+		// Base-32 uses an exclusively uppercase alphabet, per spec, but DNS is
+		// case-insensitive. Convert to uppercase to ensure well-formed names will
+		// decode.
+		if _, err := base32.StdEncoding.DecodeString(strings.ToUpper(onionDomain)); err != nil {
```

Comment:
Oh shoot, re-reading the docs while updating this comment, Golang will ignore newlines, which means you can still smuggle in invalid values.

I'll update this to first pass through a regex.

Reply:
OK, swapped to a regex, which makes the base32 encoding unnecessary. This is because the alignment (16, 56) means that the inputs are integral multiples of 40 input bits -> 8 output bytes, which means that if they match the regex, they are by definition correctly encoded.

This avoids golang allowing padding (e.g. a 12-byte input with 4 bytes of = padding) and newlines